diff --git a/call/rtp_transport_controller_send.cc b/call/rtp_transport_controller_send.cc
index d743a0bf43..2b6a3209bd 100644
--- a/call/rtp_transport_controller_send.cc
+++ b/call/rtp_transport_controller_send.cc
@@ -127,3 +127,3 @@ RtpTransportControllerSend::RtpTransportControllerSend(
           "rtp_send_controller",
-          TaskQueueFactory::Priority::NORMAL)) {
+          TaskQueueFactory::Priority::HIGH)) {
   ParseFieldTrial({&relay_bandwidth_cap_},
diff --git a/modules/pacing/task_queue_paced_sender.cc b/modules/pacing/task_queue_paced_sender.cc
index f2de9ecb64..6ee946bfb5 100644
--- a/modules/pacing/task_queue_paced_sender.cc
+++ b/modules/pacing/task_queue_paced_sender.cc
@@ -53,7 +53,7 @@ TaskQueuePacedSender::TaskQueuePacedSender(
       packet_size_(/*alpha=*/0.95),
       task_queue_(task_queue_factory->CreateTaskQueue(
           "TaskQueuePacedSender",
-          TaskQueueFactory::Priority::NORMAL)) {
+          TaskQueueFactory::Priority::HIGH)) {
   packet_size_.Apply(1, 0);
 }

diff --git a/sdk/objc/helpers/RTCDispatcher.m b/sdk/objc/helpers/RTCDispatcher.m
index 4df19bc297..408c1e0258 100644
--- a/sdk/objc/helpers/RTCDispatcher.m
+++ b/sdk/objc/helpers/RTCDispatcher.m
@@ -23,8 +23,8 @@ + (void)initialize {
         "org.webrtc.RTCDispatcherAudioSession",
-        DISPATCH_QUEUE_SERIAL);
+        DISPATCH_QUEUE_CONCURRENT);
     kCaptureSessionQueue = dispatch_queue_create(
         "org.webrtc.RTCDispatcherCaptureSession",
-        DISPATCH_QUEUE_SERIAL);
+        DISPATCH_QUEUE_CONCURRENT);
     kNetworkMonitorQueue =
-        dispatch_queue_create("org.webrtc.RTCDispatcherNetworkMonitor", DISPATCH_QUEUE_SERIAL);
+        dispatch_queue_create("org.webrtc.RTCDispatcherNetworkMonitor", DISPATCH_QUEUE_CONCURRENT);
   });
diff --git a/video/video_receive_stream2.cc b/video/video_receive_stream2.cc
index c9ec9e0123..58cb834e6c 100644
--- a/video/video_receive_stream2.cc
+++ b/video/video_receive_stream2.cc
@@ -260,3 +260,3 @@ VideoReceiveStream2::VideoReceiveStream2(
           "DecodingQueue",
-          TaskQueueFactory::Priority::HIGH)) {
+          TaskQueueFactory::Priority::LOW)) {
   RTC_LOG(LS_INFO) << "VideoReceiveStream2: " << config_.ToString();
diff --git a/video/video_stream_decoder_impl.cc b/video/video_stream_decoder_impl.cc
index b6d754e8be..2139ce8c76 100644
--- a/video/video_stream_decoder_impl.cc
+++ b/video/video_stream_decoder_impl.cc
@@ -37,6 +37,6 @@ VideoStreamDecoderImpl::VideoStreamDecoderImpl(
           "video_stream_decoder_bookkeeping_queue",
-          TaskQueueFactory::Priority::NORMAL)),
+          TaskQueueFactory::Priority::LOW)),
       decode_queue_(task_queue_factory->CreateTaskQueue(
           "video_stream_decoder_decode_queue",
-          TaskQueueFactory::Priority::NORMAL)) {
+          TaskQueueFactory::Priority::LOW)) {
   bookkeeping_queue_.PostTask([this]() {
diff --git a/modules/audio_device/audio_device_buffer.cc b/modules/audio_device/audio_device_buffer.cc
index 977045419a..fe14b3a20a 100644
--- a/modules/audio_device/audio_device_buffer.cc
+++ b/modules/audio_device/audio_device_buffer.cc
@@ -45,3 +45,3 @@ AudioDeviceBuffer::AudioDeviceBuffer(TaskQueueFactory* task_queue_factory)
           kTimerQueueName,
-          TaskQueueFactory::Priority::NORMAL)),
+          TaskQueueFactory::Priority::HIGH)),
       audio_transport_cb_(nullptr),
diff --git a/audio/channel_send.cc b/audio/channel_send.cc
index 47afc7982b..58e7051f75 100644
--- a/audio/channel_send.cc
+++ b/audio/channel_send.cc
@@ -474,3 +474,3 @@ ChannelSend::ChannelSend(
           "AudioEncoder",
-          TaskQueueFactory::Priority::NORMAL)),
+          TaskQueueFactory::Priority::HIGH)),
       fixing_timestamp_stall_(
diff --git a/audio/voip/audio_egress.cc b/audio/voip/audio_egress.cc
index 1162824c9e..e72f7078ba 100644
--- a/audio/voip/audio_egress.cc
+++ b/audio/voip/audio_egress.cc
@@ -27,3 +27,3 @@ AudioEgress::AudioEgress(RtpRtcpInterface* rtp_rtcp,
           "AudioEncoder",
-          TaskQueueFactory::Priority::NORMAL)) {
+          TaskQueueFactory::Priority::HIGH)) {
   audio_coding_->RegisterTransportCallback(this);
diff --git a/modules/async_audio_processing/async_audio_processing.cc b/modules/async_audio_processing/async_audio_processing.cc
index 9452f3bcf9..c7bfaa2667 100644
--- a/modules/async_audio_processing/async_audio_processing.cc
+++ b/modules/async_audio_processing/async_audio_processing.cc
@@ -46,3 +46,3 @@ AsyncAudioProcessing::AsyncAudioProcessing(
           "AsyncAudioProcessing",
-          TaskQueueFactory::Priority::NORMAL)) {
+          TaskQueueFactory::Priority::HIGH)) {
   frame_processor_.SetSink([this](std::unique_ptr<AudioFrame> frame) {
