diff --git a/video/video_stream_encoder.cc b/video/video_stream_encoder.cc
index 107110987b..6029554c8f 100644
--- a/video/video_stream_encoder.cc
+++ b/video/video_stream_encoder.cc
@@ -661,3 +661,3 @@ VideoStreamEncoder::VideoStreamEncoder(
           "EncoderQueue",
-          TaskQueueFactory::Priority::NORMAL)) {
+          TaskQueueFactory::Priority::HIGH)) {
   TRACE_EVENT0("webrtc", "VideoStreamEncoder::VideoStreamEncoder");
diff --git a/modules/pacing/task_queue_paced_sender.cc b/modules/pacing/task_queue_paced_sender.cc
index 709718ff16..d8f9b4c17c 100644
--- a/modules/pacing/task_queue_paced_sender.cc
+++ b/modules/pacing/task_queue_paced_sender.cc
@@ -50,7 +50,7 @@ TaskQueuePacedSender::TaskQueuePacedSender(
       is_shutdown_(false),
       task_queue_(task_queue_factory->CreateTaskQueue(
           "TaskQueuePacedSender",
-          TaskQueueFactory::Priority::NORMAL)) {}
+          TaskQueueFactory::Priority::HIGH)) {}

 TaskQueuePacedSender::~TaskQueuePacedSender() {
   // Post an immediate task to mark the queue as shutting down.
diff --git a/sdk/objc/helpers/RTCDispatcher.m b/sdk/objc/helpers/RTCDispatcher.m
index 4df19bc297..408c1e0258 100644
--- a/sdk/objc/helpers/RTCDispatcher.m
+++ b/sdk/objc/helpers/RTCDispatcher.m
@@ -21,12 +21,12 @@ + (void)initialize {
   dispatch_once(&onceToken, ^{
     kAudioSessionQueue = dispatch_queue_create(
         "org.webrtc.RTCDispatcherAudioSession",
-        DISPATCH_QUEUE_SERIAL);
+        DISPATCH_QUEUE_CONCURRENT);
     kCaptureSessionQueue = dispatch_queue_create(
         "org.webrtc.RTCDispatcherCaptureSession",
-        DISPATCH_QUEUE_SERIAL);
+        DISPATCH_QUEUE_CONCURRENT);
     kNetworkMonitorQueue =
-        dispatch_queue_create("org.webrtc.RTCDispatcherNetworkMonitor", DISPATCH_QUEUE_SERIAL);
+        dispatch_queue_create("org.webrtc.RTCDispatcherNetworkMonitor", DISPATCH_QUEUE_CONCURRENT);
   });
 }
