diff --git a/BUILD.gn b/BUILD.gn
index 314e7a6165..cc1b12d237 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -119,10 +119,10 @@ config("common_inherited_config") {
   ldflags = []
 
   if (rtc_enable_symbol_export || is_component_build) {
-    defines = [ "WEBRTC_ENABLE_SYMBOL_EXPORT" ]
+    defines += [ "WEBRTC_ENABLE_SYMBOL_EXPORT" ]
   }
   if (rtc_enable_objc_symbol_export) {
-    defines = [ "WEBRTC_ENABLE_OBJC_SYMBOL_EXPORT" ]
+    defines += [ "WEBRTC_ENABLE_OBJC_SYMBOL_EXPORT" ]
   }
 
   if (build_with_mozilla) {
