diff --git a/modules/desktop_capture/BUILD.gn b/modules/desktop_capture/BUILD.gn
index 25b92bed45..9fc003e64d 100644
--- a/modules/desktop_capture/BUILD.gn
+++ b/modules/desktop_capture/BUILD.gn
@@ -282,6 +282,7 @@ rtc_source_set("desktop_capture") {
 if (is_mac) {
   rtc_library("desktop_capture_objc") {
     visibility = [ ":desktop_capture" ]
+    cflags = [ "-Wno-deprecated-declarations" ]
     sources = [
       "mac/desktop_configuration.mm",
       "mac/desktop_frame_cgimage.h",

diff --git a/modules/audio_device/BUILD.gn b/modules/audio_device/BUILD.gn
index 5d6a1d82fc..5b440ecf1e 100644
--- a/modules/audio_device/BUILD.gn
+++ b/modules/audio_device/BUILD.gn
@@ -294,6 +294,7 @@ rtc_library("audio_device_impl") {
         ]
       }
       if (is_mac) {
+        cflags += [ "-Wno-deprecated-declarations" ]
         sources += [
           "mac/audio_device_mac.cc",
           "mac/audio_device_mac.h",

diff --git a/modules/third_party/portaudio/BUILD.gn b/modules/third_party/portaudio/BUILD.gn
index c49c544e9d..a43f8a9e6a 100644
--- a/modules/third_party/portaudio/BUILD.gn
+++ b/modules/third_party/portaudio/BUILD.gn
@@ -10,6 +10,7 @@ import("../../../webrtc.gni")

 rtc_library("mac_portaudio") {
   visibility = [ "../../audio_device:*" ]
+  cflags = [ "-Wno-deprecated-declarations" ]
   sources = [
     "pa_memorybarrier.h",
     "pa_ringbuffer.c",
